added d3 chart for opcounters

aherlihy opened the pull request:
Hi! @rueckstiess @durran @imlucas the d3 code I wrote is all in sschart.jsx. 1 thing to note - Kevin sent me a file with the HTML colors from the prototype but they look completely different on my screen. He said he'd look into it and get back to me with the correct colors.

I had to get fonts from Kevin as well, if you need them too let me know.

Thanks!

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mongodb-js/server-stats/2)

<!-- Reviewable:end -->

rueckstiess commented:
Looks really good! Amazing progress in the last days :-) 

Some comments in general: 

- If most of the `*-component.jsx` files are identical, let's only make one generic one, and pass in the `className` or whatever else they differ on as a prop (DRY). Or is there another reason why the are all separate files?
- What editor do you use? Can you set it up so that it uses eslint and the `.eslintrc` file? It will help you avoid a lot of potential issues. If you use atom (I would recommend it), I can help you set it up. If you use vim, Durran is your guy! My editor found 204 issues in the `sschart.jsx` file :-) I won't list them all here, let's get eslint working for you first.
- I would like to re-implement the stores using the reflux-state-mixin pattern that I've used in the `khaos-compass-package` template. It makes for much cleaner code IMO. Let's talk about it! Maybe we can put your code inside that template at the same time. It will help you develop faster, too.
- Many `TODO` comments. Are they still TO DO or are they done?

More comments inline.

---

Reviewed 19 of 20 files at r1.
Review status: 18 of 19 files reviewed at latest revision, 16 unresolved discussions.

---

*[package.json, line 32 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KPzx_MsHd-Yked3rnmo:-KPzx_MsHd-Yked3rnmp:b945yeh) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/package.json#L32)):*
> ```JSON
>   "dependencies": {
>     "d3": "^3.5.6",
>     "debug": "github:mongodb-js/debug#v2.2.3",
> ```

let's move THIS debug down to the devDependencies, and replace the off-the-shelf debug there.

---

*[src/components/globallock-component.jsx, line 44 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KPzzmuJf5mqmuIFOoyP:-KPzzmuJf5mqmuIFOoyQ:b-tzho8b) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/components/globallock-component.jsx#L44)):*
> ```JSX
>   componentWillUnmount() {
>     this.unsubscribeRefresh();
>     clearInterval(this.intervalId);
> ```

this.intervalID is commented out in `componentDidMount`. Need to remove it here, too? Why is it not used anymore?

---

*[src/components/globallock-component.jsx, line 86 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ--90WK2RQ7RwIN1NH:-KQ--90WK2RQ7RwIN1NI:bdjqx3c) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/components/globallock-component.jsx#L86)):*
> ```JSX
>    */
>   renderGraph() {
>     if (this.state.data && 'localTime' in this.state.data) {
> ```

what if the statement is not true? should at least `return null;` at the end.

---

*[src/components/mem-component.jsx, line 45 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ--S3Lfifi1TJi7vMT:-KQ--S3Lfifi1TJi7vMU:b-758m7s) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/components/mem-component.jsx#L45)):*
> ```JSX
>   componentWillUnmount() {
>     this.unsubscribeRefresh();
>     clearInterval(this.intervalId);
> ```

same here as in globallock-component.

---

*[src/components/mem-component.jsx, line 104 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ--ejRi-pwUglNqQvn:-KQ--ejRi-pwUglNqQvo:b-cjcf1u) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/components/mem-component.jsx#L104)):*
> ```JSX
> MemComponent.displayName = 'MemComponent';
> 
> module.exports = MemComponent;
> ```

It looks like globallock-component and mem-component are almost exactly identical. Is it just the `className` in the `renderGraph()` function that differs? If so, we shouldn't repeat all that code. Let's just make a single component that takes an additional class name as prop and re-use where possible.

---

*[src/d3/index.js, line 2 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-0WqVaflVVd8RupNe:-KQ-0WqWzRi5bY3IuPIt:buau1qm) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/index.js#L2)):*
> ```JavaScript
> module.exports = {
>   sschart: require('./sschart')
> ```

what does `sschart` stand for? `server stats chart`? It looks like the d3 code is generic and it could render all kinds of charts representing stats. I would make it more explicit and call the file `stats-chart` or `multi-line-chart` or something more descriptive. (needs to be changed everywhere though).

---

*[src/d3/sschart.jsx, line 1 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-6HzwTkfquCpql2gr:-KQ-6HzwTkfquCpql2gs:bpqupo9) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L1)):*
> ```JSX
> const d3 = require('d3');
> ```

If the file does not contain any jsx specific code, like `<Component />`, then just use the simple `.js` ending.

---

*[src/d3/sschart.jsx, line 11 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-1iTjDEyfBvSkMJtk:-KQ-1iTjDEyfBvSkMJtl:b-86eigu) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L11)):*
> ```JSX
>   var x = d3.time.scale();
>   var y = d3.scale.linear();
>   var connectionY = d3.scale.linear(); // TODO: scale for connection differently
> ```

what is the TODO here? Is there something to be done still?

---

*[src/d3/sschart.jsx, line 21 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-7XxGANA10AEkZU4t:-KQ-7XxGANA10AEkZU4u:b-9bnoj) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L21)):*
> ```JSX
>   function chart(selection) {
>     selection.each(function(data) {
>       if (!('localTime' in data) || data.localTime.length == 0) {
> ```

You should always use `===` in JavaScript unless there is a really good reason not to (`==` does some unpredictable auto-casting and will just attract bugs).

---

*[src/d3/sschart.jsx, line 22 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-1vV9EzzKcN2zWfGX:-KQ-1vV9EzzKcN2zWfGY:b-wvtvjx) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L22)):*
> ```JSX
>     selection.each(function(data) {
>       if (!('localTime' in data) || data.localTime.length == 0) {
>         return; // TODO: okay to not draw graph with bad data?
> ```

maybe draw all the things that can be drawn and that are not data related (axes, labels, etc), and then return before it gets to the data-specific parts. So invalid data would just render an empty chart.

---

*[src/d3/sschart.jsx, line 25 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-4rL04nqCUbjmVGsX:-KQ-4rL04nqCUbjmVGsY:b-v0eaq2) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L25)):*
> ```JSX
>       }
>       keys = data.operations.map(function(f) { return f.op; });
>       var margin = {top: 60, right: 30, bottom: 50, left: 40};
> ```

`margin` never changes, move to the outer scope where you define `x` and `y`. If there's anything else that doesn't depend on data or width/height, that can also be moved out. 

---

*[src/d3/sschart.jsx, line 130 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-89fQK4fioSUJvCEJ:-KQ-89fQK4fioSUJvCEK:b-qnhi1q) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L130)):*
> ```JSX
>         .x(function(d, i) { return x(data.localTime[i]); })
>         .y(function(d) { return y(d); });
>       g.selectAll('.operation').data(data.operations)
> ```

is `operation` just a catchall phrase for any kind of stat type you get from the stores? i.e., could it be operations, or disk read/writes, or memory, etc? If so, maybe choose a more descriptive variable and css class name, like `stat`. Especially since it is not just limited to this file but the stores have to use the same variable name, too. 

---

*[src/d3/sschart.jsx, line 134 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-AKbfU4bemuibZS5g:-KQ-AKbfU4bemuibZS5h:benhfap) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L134)):*
> ```JSX
>         .attr('class', 'operation')
>         .append('path')
>         .attr('class', function(d) { return 'line ' + d.op; })
> ```

To make it more future proof, I would suggest we don't use classes for the actual operations here, but use generic numbered classes. 

```
.attr('class', function(d, i) { return `line chart-color-${i}`; })
```

And in the css file, rename the op-specific classes to `chart-color-0`, `chart-color-1`. That way, we can re-use this code, or add new charts, without having to change the style files and spell out all individual operation names.

---

*[src/d3/sschart.jsx, line 135 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-8wKz-S8pC9ysjc04:-KQ-8wL-Utvt5QsUvk9N:butzoje) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L135)):*
> ```JSX
>         .append('path')
>         .attr('class', function(d) { return 'line ' + d.op; })
>         .style('fill', 'none')
> ```

We'll probably want to have all the styles in the stylesheet file instead of inline. That's true for all the `.style` calls.

---

*[src/d3/sschart.jsx, line 295 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-9jq3vpY3zjdzX2Rk:-KQ-9jq47D77iNNNOMiG:b-mktfp) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L295)):*
> ```JSX
>   return chart;
> };
> module.exports = graphfunction;
> ```

Could you add many more comments in this file? :-) Just describe a bit more what the individual blocks are doing, for example the `updateOverlay` function is pretty dense.

---

*[styles/index.css, line 63 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-BBCVQMf4oSH-KVVR:-KQ-BBCWfZZdkpPuwJBD:bedgo8l) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/styles/index.css#L63)):*
> ```CSS
>     stroke: none;
> }
> .insert, .bytesOut, .virtual, .aReads { /* teal */
> ```

lets replace all those and below classes with generic `chart-color-0`, `chart-color-1` etc so we don't have to hard-code what charts we want to use this module for (see comment above in sschart.jsx).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-:-KQ-2YVzMcBcHI9F9oCG:b-fpri3j)*
<!-- Sent from Reviewable.io -->

aherlihy commented:
Review status: 18 of 19 files reviewed at latest revision, 16 unresolved discussions.

---

*[src/components/globallock-component.jsx, line 44 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KPzzmuJf5mqmuIFOoyP:-KQ-oC8ePKFiTsG1PDdr:b-op04pf) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/components/globallock-component.jsx#L44)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> this.intervalID is commented out in `componentDidMount`. Need to remove it here, too? Why is it not used anymore?
</details>
The intervalId only needs to set once, and it's being set in opcounters-component. If this was uncommented (as well as this in mem-component/network-component) then it would trigger the serverStatus() action once for every graph, and we need it to be triggered only once for all of them. This obviously needs to move but I'm not sure where to put it. Also if we combine all the different chart components (per your other comment) it'd have to make sure that it only triggers once.

---

*[src/components/globallock-component.jsx, line 86 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ--90WK2RQ7RwIN1NH:-KQ-p65QqIYaQIcXGqtY:bnxct6n) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/components/globallock-component.jsx#L86)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> what if the statement is not true? should at least `return null;` at the end.
</details>
Yeah I wanted to ask you what the behavior should be if the data is bad - right now it just doesn't draw anything. Is returning null when there's no data a reasonable thing to do?

---

*[src/components/mem-component.jsx, line 45 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ--S3Lfifi1TJi7vMT:-KQ-pHV1WtaPnPG51jle:bc7kezw) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/components/mem-component.jsx#L45)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> same here as in globallock-component.
</details>
I should get rid of the clearInterval calls since they're not setting the interval, but see the above comment about finding a place to trigger the serverStatus action.

---

*[src/components/mem-component.jsx, line 104 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ--ejRi-pwUglNqQvn:-KQ-pS0zIEbxrUxv0SKO:b-lmh1ct) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/components/mem-component.jsx#L104)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> It looks like globallock-component and mem-component are almost exactly identical. Is it just the `className` in the `renderGraph()` function that differs? If so, we shouldn't repeat all that code. Let's just make a single component that takes an additional class name as prop and re-use where possible.
</details>
Ok, I can do that but (from the above comment) I'm not sure how to deal with triggering the action once.

---

*[src/d3/index.js, line 2 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-0WqVaflVVd8RupNe:-KQ-pn525w-Wa64GzQ8r:b-lnvqpq) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/index.js#L2)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> what does `sschart` stand for? `server stats chart`? It looks like the d3 code is generic and it could render all kinds of charts representing stats. I would make it more explicit and call the file `stats-chart` or `multi-line-chart` or something more descriptive. (needs to be changed everywhere though).
</details>
yeah server stats chart - I can rename it to stats-chart. 

---

*[src/d3/sschart.jsx, line 1 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-6HzwTkfquCpql2gr:-KQ-px5Hl5wAdN0XCJze:b-bpz02z) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L1)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> If the file does not contain any jsx specific code, like `<Component />`, then just use the simple `.js` ending.
</details>
so I should rename this file to .js? Does anything else (other than import statements) need to change?

---

*[src/d3/sschart.jsx, line 11 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-1iTjDEyfBvSkMJtk:-KQ-q2kGKAirmKQ08CN6:bqytjd) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L11)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> what is the TODO here? Is there something to be done still?
</details>
Yeah if you notice on the prototype (https://mongodb.invisionapp.com/share/KG86GCC8E#/screens/178973271) the connection line is scaled completely differently than the other two (there's a label on the rightmost border). Personally I think that's really confusing to have different scales for different lines, but that's the way the prototype works. I wanted to ask you if there was a good way to implement that.

---

*[src/d3/sschart.jsx, line 22 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-1vV9EzzKcN2zWfGX:-KQ-qhtHuYLDX78aBN7s:b32w0x) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L22)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> maybe draw all the things that can be drawn and that are not data related (axes, labels, etc), and then return before it gets to the data-specific parts. So invalid data would just render an empty chart.
</details>
OK

---

*[src/d3/sschart.jsx, line 25 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-4rL04nqCUbjmVGsX:-KQ-rMryWGO4g9KCInIl:b32w0x) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L25)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> `margin` never changes, move to the outer scope where you define `x` and `y`. If there's anything else that doesn't depend on data or width/height, that can also be moved out. 
</details>
OK

---

*[src/d3/sschart.jsx, line 130 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-89fQK4fioSUJvCEJ:-KQ-rRiPxk7oXu11iBHR:b32w0x) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L130)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> is `operation` just a catchall phrase for any kind of stat type you get from the stores? i.e., could it be operations, or disk read/writes, or memory, etc? If so, maybe choose a more descriptive variable and css class name, like `stat`. Especially since it is not just limited to this file but the stores have to use the same variable name, too. 
</details>
OK

---

*[src/d3/sschart.jsx, line 134 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-AKbfU4bemuibZS5g:-KQ-rZveYsBc29tsaZVN:b-opopns) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L134)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> To make it more future proof, I would suggest we don't use classes for the actual operations here, but use generic numbered classes. 
> 
> ```
> .attr('class', function(d, i) { return `line chart-color-${i}`; })
> ```
> 
> And in the css file, rename the op-specific classes to `chart-color-0`, `chart-color-1`. That way, we can re-use this code, or add new charts, without having to change the style files and spell out all individual operation names.
</details>
Ok, but if I'm copying the colors from the prototype I don't think they are always in the same order as the keys in the legend underneath. If I do it by number then it won't match the prototype exactly.

---

*[src/d3/sschart.jsx, line 135 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-8wKz-S8pC9ysjc04:-KQ-rgl8W3J1e58imxyo:b-mwn26m) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L135)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> We'll probably want to have all the styles in the stylesheet file instead of inline. That's true for all the `.style` calls.
</details>
I thought this style needs to go here because it goes on/off depending on mouseover/etc?

---

*[src/d3/sschart.jsx, line 295 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-9jq3vpY3zjdzX2Rk:-KQ-rnZEQuacpa-nW-hX:b-m8quha) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L295)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> Could you add many more comments in this file? :-) Just describe a bit more what the individual blocks are doing, for example the `updateOverlay` function is pretty dense.
</details>
Yes, sorry  comments are always good :)

---

*[styles/index.css, line 63 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-BBCVQMf4oSH-KVVR:-KQ-sFCCcgqf2RE1qePd:b-j82po2) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/styles/index.css#L63)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> lets replace all those and below classes with generic `chart-color-0`, `chart-color-1` etc so we don't have to hard-code what charts we want to use this module for (see comment above in sschart.jsx).
</details>
Sure, but see above comment about color order.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mongodb-js/server-stats/2)*
<!-- Sent from Reviewable.io -->

rueckstiess commented:
Review status: 18 of 19 files reviewed at latest revision, 11 unresolved discussions.

---

*[src/components/globallock-component.jsx, line 44 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KPzzmuJf5mqmuIFOoyP:-KQ-zaEENnu4Aa05V-d4:bbc3mrs) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/components/globallock-component.jsx#L44)):*
<details><summary><i>Previously, aherlihy wrote…</i></summary>
> The intervalId only needs to set once, and it's being set in opcounters-component. If this was uncommented (as well as this in mem-component/network-component) then it would trigger the serverStatus() action once for every graph, and we need it to be triggered only once for all of them. This obviously needs to move but I'm not sure where to put it. Also if we combine all the different chart components (per your other comment) it'd have to make sure that it only triggers once.
</details>
Ah, got it, it's the n-second interval timer. It should go into the server-stats store. If I see it correctly, your other stores are being notified of any server-stats store changes, so the server-stats store is the initiator. We can do that as part of the store restructuring. 

---

*[src/components/globallock-component.jsx, line 86 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ--90WK2RQ7RwIN1NH:-KQ-zwOpJaruhQ7MxcT-:b-g6nq1a) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/components/globallock-component.jsx#L86)):*
<details><summary><i>Previously, aherlihy wrote…</i></summary>
> Yeah I wanted to ask you what the behavior should be if the data is bad - right now it just doesn't draw anything. Is returning null when there's no data a reasonable thing to do?
</details>
At least for now, let's return `null`. React understands `null` components: https://facebook.github.io/react/blog/2014/07/17/react-v0.11.html#rendering-to-null

We should ask the designers about what they had in mind in terms of error handling though. We need to give the user some feedback. 

---

*[src/components/mem-component.jsx, line 45 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ--S3Lfifi1TJi7vMT:-KQ0-X0dHoHd6WxrwgrO:b9d6q8l) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/components/mem-component.jsx#L45)):*
<details><summary><i>Previously, aherlihy wrote…</i></summary>
> I should get rid of the clearInterval calls since they're not setting the interval, but see the above comment about finding a place to trigger the serverStatus action.
</details>
Yes, let's remove all setInterval / clearInterval code from the components and move that to the server-stats store. 

---

*[src/d3/sschart.jsx, line 1 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-6HzwTkfquCpql2gr:-KQ0-odcFjWsLgEviLig:b-or45h7) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L1)):*
<details><summary><i>Previously, aherlihy wrote…</i></summary>
> so I should rename this file to .js? Does anything else (other than import statements) need to change?
</details>
no, that should be it. if it's `.js` you can require it without extension, e.g.:
```
const statsChart = require('./stats-chart')
```

---

*[src/d3/sschart.jsx, line 11 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-1iTjDEyfBvSkMJtk:-KQ007Mj7mXO8y0LGsOB:b87z34i) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L11)):*
<details><summary><i>Previously, aherlihy wrote…</i></summary>
> Yeah if you notice on the prototype (https://mongodb.invisionapp.com/share/KG86GCC8E#/screens/178973271) the connection line is scaled completely differently than the other two (there's a label on the rightmost border). Personally I think that's really confusing to have different scales for different lines, but that's the way the prototype works. I wanted to ask you if there was a good way to implement that.
</details>
ok. pretty much exactly the same why as you did with the first y-scale / axis. you need to set the domain and range correctly, call the axis object, etc, and when you draw the connections line, just use `connectionY()` instead of `y()` where you generate the path line.

---

*[src/d3/sschart.jsx, line 134 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-AKbfU4bemuibZS5g:-KQ00fXZqePZ_kXiL95K:b7ggnxz) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L134)):*
let's confirm with designers. we need to know:

>Can you give us an ordering of the colors we can use for all charts independent of the chart type? (ideally in order of most contrasting. d3 does that, too).

---

*[src/d3/sschart.jsx, line 135 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-8wKz-S8pC9ysjc04:-KQ02_XIrNzFzcVKAFN7:b-lb4agy) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L135)):*
<details><summary><i>Previously, aherlihy wrote…</i></summary>
> I thought this style needs to go here because it goes on/off depending on mouseover/etc?
</details>
Ok, for the styles that need to change dynamically (e.g. for mouseover), they can stay. If it is something constant, let's move to the stylesheet.
 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mongodb-js/server-stats/2)*
<!-- Sent from Reviewable.io -->

aherlihy commented:
- Done, put all the component files in one.
- eslintrc working sort of, it's at least scanning the d3 since it's not jsx anymore. Let me know if you figure out a way to get eslint to run on .jsx.
- I never got the template version working, I'm going to need your help figuring it out. As is, the lines aren't being drawn but everything else is. I've looked at the data being received by the charts and it looks correct. 100% out of ideas, I have absolutely no idea why it's not working. 

---

Review status: 18 of 19 files reviewed at latest revision, 11 unresolved discussions.

---

*[package.json, line 32 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KPzx_MsHd-Yked3rnmo:-KQ1DGYVbh-pFIoeqHIr:b-896fix) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/package.json#L32)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> let's move THIS debug down to the devDependencies, and replace the off-the-shelf debug there.
</details>
Done.

---

*[src/components/globallock-component.jsx, line 44 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KPzzmuJf5mqmuIFOoyP:-KQ1ELdPAZA77ecwFxpk:bjwahcf) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/components/globallock-component.jsx#L44)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> Ah, got it, it's the n-second interval timer. It should go into the server-stats store. If I see it correctly, your other stores are being notified of any server-stats store changes, so the server-stats store is the initiator. We can do that as part of the store restructuring. 
</details>
moved it into ServerStatsStore.init for now, but when its integrated into compass it'll have an overarching server-status-component that will do it.

---

*[src/components/globallock-component.jsx, line 86 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ--90WK2RQ7RwIN1NH:-KQ-u7g7raK9pMl3g685:b-uukpbs) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/components/globallock-component.jsx#L86)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> At least for now, let's return `null`. React understands `null` components: https://facebook.github.io/react/blog/2014/07/17/react-v0.11.html#rendering-to-null
> 
> We should ask the designers about what they had in mind in terms of error handling though. We need to give the user some feedback. 
</details>
Sure, although the # keys and key names come from data, as well as the title, so the only thing we can draw is that background grid. Which may not be particularly useful to the user. I've changed it so it displays a "bad data" error message, we can get more specifics from the designers later.

---

*[src/components/mem-component.jsx, line 45 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ--S3Lfifi1TJi7vMT:-KQ1Q1-3AcPClJrtWJjp:b-896fix) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/components/mem-component.jsx#L45)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> Yes, let's remove all setInterval / clearInterval code from the components and move that to the server-stats store. 
</details>
Done.

---

*[src/components/mem-component.jsx, line 104 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ--ejRi-pwUglNqQvn:-KQ1br9JF_H9kfTbnQ4a:b-896fix) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/components/mem-component.jsx#L104)):*
<details><summary><i>Previously, aherlihy wrote…</i></summary>
> Ok, I can do that but (from the above comment) I'm not sure how to deal with triggering the action once.
</details>
Done.

---

*[src/d3/index.js, line 2 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-0WqVaflVVd8RupNe:-KQ1Q6x-M420eHhHh2SZ:b-896fix) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/index.js#L2)):*
<details><summary><i>Previously, aherlihy wrote…</i></summary>
> yeah server stats chart - I can rename it to stats-chart. 
</details>
Done.

---

*[src/d3/sschart.jsx, line 1 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-6HzwTkfquCpql2gr:-KQ1QA5tg6ntoN1Ssa2-:b-896fix) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L1)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> no, that should be it. if it's `.js` you can require it without extension, e.g.:
> ```
> const statsChart = require('./stats-chart')
> ```
</details>
Done.

---

*[src/d3/sschart.jsx, line 11 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-1iTjDEyfBvSkMJtk:-KQ1bshDOYYMeZ0jvzCM:b-3pqb0j) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L11)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> ok. pretty much exactly the same why as you did with the first y-scale / axis. you need to set the domain and range correctly, call the axis object, etc, and when you draw the connections line, just use `connectionY()` instead of `y()` where you generate the path line.
</details>
This will probably wait for a V2.

---

*[src/d3/sschart.jsx, line 21 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-7XxGANA10AEkZU4t:-KQ1Q_DbslZzsBSaiUx_:b-896fix) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L21)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> You should always use `===` in JavaScript unless there is a really good reason not to (`==` does some unpredictable auto-casting and will just attract bugs).
</details>
Done.

---

*[src/d3/sschart.jsx, line 134 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-AKbfU4bemuibZS5g:-KQ1auYKJ7LRddzQWL_R:b-z1wtjd) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L134)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> let's confirm with designers. we need to know:
> 
> >Can you give us an ordering of the colors we can use for all charts independent of the chart type? (ideally in order of most contrasting. d3 does that, too).
> 
</details>
Had a conversation with Kevin, he says he'll talk to the other designers about it and figure out if it's possible. They wanted the colors to correspond to types of operations/etc, i.e. delete=red and queued writes=red. 

---

*[src/d3/sschart.jsx, line 135 \[r1\]](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-KQ-8wKz-S8pC9ysjc04:-KQ1QaRrCl5nocrxFnFI:b-42p4kf) ([raw file](https://github.com/mongodb-js/server-stats/blob/cc9eb7f29fc30c310099d9594064a2c1c428261a/src/d3/sschart.jsx#L135)):*
<details><summary><i>Previously, rueckstiess (Thomas Rückstieß) wrote…</i></summary>
> Ok, for the styles that need to change dynamically (e.g. for mouseover), they can stay. If it is something constant, let's move to the stylesheet.
>  
</details>
I don't think there are any styles that are constant, do you see any?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mongodb-js/server-stats/2#-:-KQ1b9QD15FEgmndxTW4:bnv5765)*
<!-- Sent from Reviewable.io -->

aherlihy commented:
Implemented in Compass repo.
